server/protocols/jmap-draft/src/main/java/org/apache/james/jmap/draft/methods/MIMEMessageConverter.java

```diff
@@ -220,6 +216,46 @@ private Multipart createMultipart(CreationMessage newMessage, ImmutableList<Mess
         }
     }
 
+    private Multipart createMultipartWithAttachments(CreationMessage newMessage, ImmutableList<MessageAttachment> messageAttachments) throws IOException {
+        MultipartBuilder mixedMultipartBuilder = MultipartBuilder.create(MIXED_SUB_TYPE);
+        Map<Boolean, List<MessageAttachment>> distinguishedAttachments = distinguishAttachments(messageAttachments);
+        List<MessageAttachment> inlineAttachments = Optional.ofNullable(distinguishedAttachments.get(true))
+            .orElse(ImmutableList.of());
+        List<MessageAttachment> normalAttachments = Optional.ofNullable(distinguishedAttachments.get(false))
+            .orElse(ImmutableList.of());
+
+        if (!inlineAttachments.isEmpty()) {
+            MultipartBuilder relatedMultipart = MultipartBuilder.create(RELATED_SUB_TYPE);
+            addBody(newMessage, relatedMultipart);
+            addAttachments(inlineAttachments, relatedMultipart);
+
+            mixedMultipartBuilder
+                .addBodyPart(Message.Builder.of()
+                    .setBody(relatedMultipart.build())
+                    .build());
+        }
+
+        if (!normalAttachments.isEmpty()) {
+            addBody(newMessage, mixedMultipartBuilder);
+            addAttachments(normalAttachments, mixedMultipartBuilder);
+        }
+
+        return mixedMultipartBuilder.build();
+    }
+
+    private MultipartBuilder addAttachments(List<MessageAttachment> messageAttachments,
+                                            MultipartBuilder multipartBuilder) {
+        Consumer<MessageAttachment> addAttachment = addAttachment(multipartBuilder);
+        messageAttachments.forEach(addAttachment);
+
+        return multipartBuilder;
+    }
+
+    private Map<Boolean, List<MessageAttachment>> distinguishAttachments(List<MessageAttachment> messageAttachments) {
```

Comment:
This map with a boolean as key does not express the intent well. Can we find a better way to separate attachments on wether they are  inlined?

Iterating two time the collection seems like a better  option.

Reply:
or we can return a pair? but maybe that's not really explicit neither

Reply:
okay, I took the filter